File: TicTacToe.py
# Tic Tac Toe Game in Python

# Function to print the board
def print_board(board):
    print("\n")
    for row in board:
        print(" | ".join(row))
        print("-" * 5)
    print("\n")

# Function to check if a player wins
def check_winner(board, player):
    # Rows
    for row in board:
        if all(cell == player for cell in row):
            return True
    # Columns
    for col in range(3):
        if all(board[row][col] == player for row in range(3)):
            return True
    # Diagonals
    if all(board[i][i] == player for i in range(3)):
        return True
    if all(board[i][2-i] == player for i in range(3)):
        return True
    return False

# Function to check draw
def is_draw(board):
    return all(cell != " " for row in board for cell in row)

# Main Game Loop
def play_game():
    board = [[" " for _ in range(3)] for _ in range(3)]
    current_player = "X"

    print("🎮 Welcome to Tic Tac Toe!")
    
    while True:
        print_board(board)
        print(f"Player {current_player}'s turn")

        try:
            row = int(input("Enter row (0-2): "))
            col = int(input("Enter column (0-2): "))
        except ValueError:
            print("❌ Invalid input! Please enter numbers only.")
            continue

        if row not in [0,1,2] or col not in [0,1,2]:
            print("❌ Position out of range! Use 0, 1, or 2.")
            continue

        if board[row][col] != " ":
            print("❌ Cell already taken! Choose another.")
            continue

        # Place move
        board[row][col] = current_player

        # Check winner
        if check_winner(board, current_player):
            print_board(board)
            print(f"🎉 Player {current_player} wins! 🎉")
            break

        # Check draw
        if is_draw(board):
            print_board(board)
            print("🤝 It's a draw!")
            break

        # Switch turn
        current_player = "O" if current_player == "X" else "X"

# Run game
play_game()

File: Tictactoe.java
import java.util.Scanner;

public class TicTacToe {

    static char[][] board = {
        { ' ', ' ', ' ' },
        { ' ', ' ', ' ' },
        { ' ', ' ', ' ' }
    };

    public static void printBoard() {
        System.out.println("\n");
        for (int i = 0; i < 3; i++) {
            System.out.println(" " + board[i][0] + " | " + board[i][1] + " | " + board[i][2]);
            if (i < 2) {
                System.out.println("---+---+---");
            }
        }
        System.out.println("\n");
    }

    public static boolean checkWinner(char player) {
        // Rows
        for (int i = 0; i < 3; i++) {
            if (board[i][0] == player && board[i][1] == player && board[i][2] == player)
                return true;
        }
        // Columns
        for (int j = 0; j < 3; j++) {
            if (board[0][j] == player && board[1][j] == player && board[2][j] == player)
                return true;
        }
        // Diagonals
        if (board[0][0] == player && board[1][1] == player && board[2][2] == player)
            return true;
        if (board[0][2] == player && board[1][1] == player && board[2][0] == player)
            return true;

        return false;
    }

    public static boolean isDraw() {
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (board[i][j] == ' ')
                    return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        char currentPlayer = 'X';
        boolean gameOver = false;

        System.out.println("🎮 Welcome to Tic Tac Toe!");
        printBoard();

        while (!gameOver) {
            System.out.println("Player " + currentPlayer + "'s turn.");
            System.out.print("Enter row (0,1,2): ");
            int row = sc.nextInt();
            System.out.print("Enter column (0,1,2): ");
            int col = sc.nextInt();

            if (row < 0 || row > 2 || col < 0 || col > 2) {
                System.out.println("❌ Invalid position! Try again.");
                continue;
            }

            if (board[row][col] != ' ') {
                System.out.println("❌ Cell already taken! Try again.");
                continue;
            }

            board[row][col] = currentPlayer;
            printBoard();

            if (checkWinner(currentPlayer)) {
                System.out.println("🎉 Player " + currentPlayer + " wins! 🎉");
                gameOver = true;
            } else if (isDraw()) {
                System.out.println("🤝 It's a draw!");
                gameOver = true;
            } else {
                currentPlayer = (currentPlayer == 'X') ? 'O' : 'X';
            }
        }

        sc.close();
    }
}
